### ISSUE TITLE
Method for determining prediction uncertainty?

### ISSUE THREAD

**Issue by rob-cohen** [catboost/catboost#954]
Problem: Is there a recommended method for determining prediction uncertainty?

I'd like to determine some kind of probability distribution for each predicted value.  For example, what is the standard error?  If it's standard error, I'd expect the value to be computed similar to MSE = (STD of locally contributing samples)/sqrt(number of locally contributing samples).

In the case of regression, I'm referring to the uncertainty of the RawFormulaVal value.
In the case of classification, I'm referring to the uncertainty of the Probaiblity value.

I've seen some brief discussion of dropping trees or using quantiles to determine MSE.

In the case of dropping trees, wouldn't that give a result that is tied to the learning rate, rather than the MSE?  Can I divide the tree contribution by the learning rate to get a calibrated residual and then compute the STD of the residuals?  Would that give me MSE?

It seems that quantile regression can be used to determine the local STD of the locally contributing training data, but then how would I transform that into uncertainty of the prediction (i.e. MSE)?  It seems like that would need some kind of weighted measure of the number of unique contributing samples.

In the case of classification, would it make more sense to somehow get a weighted measure of the number of positive and negative unique contributing samples, and then use that for a beta distribution?  Or is there a probability MSE approach that you'd suggest?

Thanks for your advice!

**Comment by rcohendw** [comment]
I tried using the raw residuals and dividing by learning rate, but I wasn't getting accurate values.  I think the problem is that residuals have been pre-scaled by a shrinking learning rate, which I'm not sure how to compute.

I'm currently using bootstrap subsampled models to provide a reference uncertainty value.  That appears to be giving me accurate numbers, but unfortunately, it requires a large number of training passes and dramatically increases storage requirements.

Since the catboost trees are already built with subsampling, I can't help but feel that this is redundant with information that's already there.  That said, even if I was able to access/compute the per-tree learning rate values, I don't really know if this is a good approach.

**Comment by david-waterworth** [comment]
Is this not something you could use cross-validation for? i.e. calculate the standard error over k-folds?

**Comment by rcohendw** [comment]
> Is this not something you could use cross-validation for? i.e. calculate the standard error over k-folds?

I appreciate the suggestion!

That's essentially what I am doing, except using bootstrap samples, rather than k-fold samples.  I read somewhere that bootstrapping was more accurate for calculating standard error, although maybe k-fold would be better stratified, for better accuracy with fewer samples.  Do you know?

I was hoping to find a more computationally efficient solution that wouldn't require a lot of passes.  Perhaps k-fold the best option?

To be clear, I want a std error value on each sample, not the total std error for all samples.  I suppose k-fold could be used either way, though.  Does the built in k-fold support std error for every value?  I wonder if that could be more efficient than running Catboost for each fold?

**Comment by rcohendw** [comment]
> Is this not something you could use cross-validation for? i.e. calculate the standard error over k-folds?

If I understand you correctly, you are suggesting cross-validation over k-folds, in order to compute the accuracy of the prediction.  i.e. standard error

Treating the prediction as a mean, I am looking for the Mean Standard Error.  The accuracy of the mean, not the prediction itself.

The difference is particularly clear with noisy data.  In that case, the standard error of the prediction may be very high, but the error between the prediction and the true mean value would be expected to be very low for lots of data.

**Comment by kizill** [comment]
We've finally made some progress with Data Uncertainty prediction for regression models! Check out our new `RMSEWithUncertainty` loss function, released in v0.24 . Model prediction have 2 elements - first is regression estimation and second is esstimation of model uncertainty due to data (NGBoost like approach). We are also currently working on total model uncertainty prediction with virtual ansembles, see our research team paper https://arxiv.org/pdf/2006.10562.pdf for more details

**Comment by kmedved** [comment]
That's very exciting @kizill - is there any documentation of how the `RMSEWithUncertainty` loss works exactly? You say it's an ngboost-like approach - is it literally via a natural gradient the same as ngboost, or something else?

A few other questions:

1) How do we interpret the estimate of model uncertainty? Is it a standard deviation or something else? (e.g, is the 2nd value returned after a `.predict()` call there the estimated standard deviation of uncertainty due to the data?). Trying to understand how to apply this.

2) Related to the above, is there any ability to, or plans to let the user specify the distribution for the uncertainty estimate? (e.g., normal distribution, lognormal, skew-T, etc...).

3) How do the results compare with RMSE regression? I ask because I immediately threw together a [Colab notebook to test this here](https://colab.research.google.com/drive/1Ajdl2G8dCaUrPH3gMclyrM5ZWLs4vPFu?usp=sharing), and as you can see, with zero tuning, the results are slightly different (even with the same random seed). Is that to be expected?.

4) I see it doesn't work on GPU - any hope that this can be added in a near release, or is that unlikely? (Obviously this has a big impact on tuning times).

Thanks again - very excited to try this out further. This is a huge step forward for GBDT models.

**Comment by MotoRZR** [comment]
Also interested in this.  Excellent addition to the library.  Would need to know how it works under the hood to satisfy stakeholders.

**Comment by lukeb88** [comment]
I am very interested in this new features. I seem to notice, however, that the official documentation has not been updated yet, am I wrong?

**Comment by oizin** [comment]
Thanks catboost team, great package.

I also wanted to mention interest in @kmedved's question: 
1. How do we interpret the estimate of model uncertainty? Is it a standard deviation or something else? (e.g, is the 2nd value returned after a .predict() call there the estimated standard deviation of uncertainty due to the data?). Trying to understand how to apply this.

In the recently uploaded [uncertainty_regression](https://github.com/catboost/catboost/blob/81b671024e93770b77a17e21981b6d50bc88d52b/catboost/tutorials/uncertainty/uncertainty_regression.ipynb) example the output of the second column from a prediction call with 'RMSEWithUncertainty' loss is called the variance. But in my own simple experiment simply predicting normally distributed noise it appears to be the standard deviation?

**Comment by DeSoham** [comment]
@oizin  They changed standard deviation to variance.

**Comment by rcohendw** [comment]
> @oizin They changed standard deviation to variance.

Is it the variance of the local samples or the uncertainty of the mean?
If the number of samples increases with the same distribution, will the uncertainty decrease?

**Comment by Evgueni-Petrov-aka-espetrov** [comment]
I think that we have answered the questions above about RMSEWithUncertainty -- just follow the links below.
https://catboost.ai/docs/references/uncertainty.html
https://catboost.ai/docs/concepts/loss-functions-regression.html

**Comment by samshipengs** [comment]
@kmedved Did you get an answer to your third question:
> 3. How do the results compare with RMSE regression? I ask because I immediately threw together a [Colab notebook to test this here](https://colab.research.google.com/drive/1Ajdl2G8dCaUrPH3gMclyrM5ZWLs4vPFu?usp=sharing), and as you can see, with zero tuning, the results are slightly different (even with the same random seed). Is that to be expected?

@Evgueni-Petrov-aka-espetrov Thanks for sharing the links, but are there any experiments comparing the performance (e.g. rmse or mae result) on `loss function = RMSE` vs `loss function = RMSEWithUncertainty`? In my current project, it seems the model trained with `RMSEWithUncertainty` (using the mean prediction) always has slightly higher `rmse` and `mae` than the one trained with just regular `RMSE`.

Does anyone have the same experience or some insights into why `RMSEWithUncertainty` might perform worse than `RMSE` (in terms of the mean prediction performance)?

**Comment by kmedved** [comment]
I don't really understand the details, although I think the answer for why `RMSEWithUncertainty` has higher `RMSE` or `MAE` scores is because it's being trained to minimize logloss, not RMSE/MAE. This is how ngboost works, and I suspect `RMSEWithUncertainty` is similar.

**Comment by samshipengs** [comment]
@kmedved thanks for sharing, have you happened to compare the performance using `RMSEWithUncertainty` vs other methods for example quantile regression? I wonder if quantile regression would give a similar uncertainty intervals while not sacrificing RMSE/MAE scores.

**Comment by executioner-jesse317** [comment]
Answer’s in place



https://lnkr.fun/049ef1c
